sha2/src/lib.rs

```diff
@@ -13,23 +13,37 @@
 #[cfg(not(any(target_arch = "x86_64", target_arch = "x86", target_arch = "aarch64")))]
 compile_error!("crate can only be used on x86, x86-64 and aarch64 architectures");
 
+cpufeatures::new!(cpuid_avx2, "avx2");
+
 #[link(name = "sha256", kind = "static")]
+#[allow(dead_code)]
 extern "C" {
     fn sha256_compress(state: &mut [u32; 8], block: &[u8; 64]);
+    fn sha256_transform_rorx(state: &mut [u32; 8], block: *const [u8; 64], num_blocks: u64);
```

Comment:
You forgot to change `num_blocks` to `usize` here. Also we probably should change `sha256_compress` to explicit pointer and length as well (same for `sha512_compress`). IIRC memory layout of slices is not guaranteed.

Reply:
It seems like it's [guaranteed](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/layout/arrays-and-slices.md#layout-of-rust-slices)

Reply:
Your link talks about layout of slice itself (i.e. about how elements of a slice a stored in memory). In this context it's more about ABI guarantees, i.e. I don't think it's currently guaranteed that `val: &[u8; 16]` is equivalent to `val_ptr: *const [u8; 16], len: usize` when used in `extern "C" fn`s.